fortitude/src/rules/correctness/error_handling.rs

```diff
@@ -0,0 +1,94 @@
+/// fails. This is often the desired behaviour, but to provide for cases in
+/// which the user wants to handle allocation errors gracefully, they may
+/// optionally check the status of an `allocate` statement by passing a variable
+/// to the `stat` argument:
+///
+/// ```f90
+/// allocate (x(BIG_INT), stat=status)
+/// if (status /= 0) then
+///   call handle_error(status)
+/// end if
+/// ```
+///
+/// However, if the `stat` variable is not checked, the program will continue to
+/// run despite the allocation failure, which can lead to undefined behaviour.
+#[derive(ViolationMetadata)]
+pub(crate) struct UncheckedAllocateStat {
+    name: String,
+}
+
+impl Violation for UncheckedAllocateStat {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Self { name } = self;
+        format!("Allocate 'stat' argument '{name}' is not checked")
+    }
+}
+
+impl AstRule for UncheckedAllocateStat {
+    fn check(_settings: &Settings, node: &Node, source: &SourceFile) -> Option<Vec<Diagnostic>> {
+        let src = source.source_text();
+
+        // Find a 'stat' argument in the allocate statement
+        let stat_node = node.named_children(&mut node.walk()).find(|child| {
+            child.kind() == "keyword_argument"
+                && child
+                    .child_by_field_name("name")
+                    .is_some_and(|n| n.to_text(src) == Some("stat"))
+        })?;
+
+        let name = stat_node
+            .child_by_field_name("value")?
+            .to_text(src)?
+            .to_string();
+
+        // Check if the 'stat' variable is checked.
+        // We do this the naive way: scan all siblings of the allocate statement
+        // and their descendants to find an instance of the variable being used
+        // somewhere. If we reach the end of the siblings without finding it, or
+        // we find it used in another allocate statement first, raise a violation.
+
+        // TODO We need a .next_siblings() iterator!
+        let mut sibling = *node;
+        'outer: while let Some(node) = sibling.next_sibling() {
+            sibling = node;
+
+            for descendant in once(sibling).chain(sibling.named_descendants()) {
+                if descendant.kind() == "allocate_statement" {
+                    // Found another allocate before checking.
+                    // TODO check that the same variable is used.
+                    break 'outer;
+                }
+                if descendant.kind() == "identifier" && descendant.to_text(src) == Some(&name) {
+                    // Found the variable, so stop checking.
+                    // Assume the user is doing something sensible with it!
+                    return None;
+                }
+            }
+        }
+        some_vec!(Diagnostic::from_node(Self { name }, &stat_node))
+
+        // TODO Also check that anything allocated in the failed statement is not used
+        // until after the variable is checked.
```

Comment:
This is almost a separate rule? `use-before-stat-check` or something. There's an argument for doing it in the same rule. Certainly it would be best to check them both at the same time -- another rule where the ruff style visitor would be useful!

Reply:
Yes, I think it should be an extra rule.